pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go

```diff
@@ -0,0 +1,489 @@
+				if !errors.IsTimeout(err) {
+					return fmt.Errorf("YurtAppDaemon[%s/%s] templatetype %s create workload by nodepool %s error: %s",
+						instance.GetNamespace(), instance.GetName(), templateType, nodepoolName, err.Error())
+				}
+			}
+			klog.Infof("YurtAppDaemon[%s/%s] templatetype %s create workload by nodepool %s success",
+				instance.GetNamespace(), instance.GetName(), templateType, nodepoolName)
+			return nil
+		})
+		if createdErr == nil {
+			r.recorder.Eventf(instance.DeepCopy(), corev1.EventTypeNormal, fmt.Sprintf("Successful %s", eventTypeWorkloadsCreated), "Create %d Workload type(%s)", createdNum, templateType)
+		} else {
+			errs = append(errs, createdErr)
+		}
+	}
+
+	// manage deleting
+	if len(needDeleted) > 0 {
+		var deleteErrs []error
+		// need deleted
+		for _, d := range needDeleted {
+			if err := r.controls[templateType].DeleteWorkload(instance, d); err != nil {
+				deleteErrs = append(deleteErrs, fmt.Errorf("YurtAppDaemon[%s/%s] delete workload[%s/%s/%s] error %v",
+					instance.GetNamespace(), instance.GetName(), templateType, d.Namespace, d.Name, err))
+			}
+		}
+		if len(deleteErrs) > 0 {
+			errs = append(errs, deleteErrs...)
+		} else {
+			r.recorder.Eventf(instance.DeepCopy(), corev1.EventTypeNormal, fmt.Sprintf("Successful %s", eventTypeWorkloadsDeleted), "Delete %d Workload type(%s)", len(needDeleted), templateType)
+		}
+	}
+
+	return len(needCreate) > 0 || len(needDeleted) > 0, utilerrors.NewAggregate(errs)
+}
+
+func (r *ReconcileYurtAppDaemon) classifyWorkloads(instance *unitv1alpha1.YurtAppDaemon, currentNodepoolToWorkload map[string]*workloadcontroller.Workload,
+	allNameToNodePools map[string]unitv1alpha1.NodePool, expectedRevision string) (needDeleted, needUpdate []*workloadcontroller.Workload,
+	needCreate []string) {
+
+	for npName, load := range currentNodepoolToWorkload {
+		find := false
+		for vnp, np := range allNameToNodePools {
+			if vnp == npName {
+				find = true
+				match := true
+				// judge workload NodeSelector
+				if !reflect.DeepEqual(load.GetNodeSelector(), workloadcontroller.CreateNodeSelectorByNodepoolName(npName)) {
+					match = false
+				}
+				// judge workload whether toleration all taints
+				match = IsTolerationsAllTaints(load.GetToleration(), np.Spec.Taints)
+
+				// judge revision
+				if load.GetRevision() != expectedRevision {
+					match = false
+				}
+
+				if !match {
+					klog.V(4).Infof("YurtAppDaemon[%s/%s] need update [%s/%s/%s]", instance.GetNamespace(),
+						instance.GetName(), load.GetKind(), load.Namespace, load.Name)
+					needUpdate = append(needUpdate, load)
+				}
+
+				break
+			}
+		}
+		if !find {
+			needDeleted = append(needDeleted, load)
+			klog.V(4).Infof("YurtAppDaemon[%s/%s] need delete [%s/%s/%s]", instance.GetNamespace(),
+				instance.GetName(), load.GetKind(), load.Namespace, load.Name)
+		}
+	}
+
+	for vnp, _ := range allNameToNodePools {
+		find := false
+		for np, _ := range currentNodepoolToWorkload {
```

Comment:
the same as above comment, use `currentNodepoolToWorkload[vnp]`

Reply:
done